core/store/src/trie/trie_storage.rs

```diff
@@ -184,30 +245,90 @@ impl TrieCachingStorage {
     fn inc_counter(&self) {
         self.counter.set(self.counter.get() + 1);
     }
+
+    /// Get position of the given key in the cache.
+    pub(crate) fn get_cache_position(&self, key: &CryptoHash) -> CachePosition {
+        match self.chunk_cache.borrow_mut().get(key) {
+            Some(value) => CachePosition::ChunkCache(value.clone()),
+            None => match self.shard_cache.get(key) {
+                Some(value) => CachePosition::ShardCache(value.clone()),
+                None => CachePosition::None,
+            },
+        }
+    }
+
+    /// If we are in the caching chunk state, put item to the chunk cache.
+    fn try_put_to_chunk_cache(&self, key: CryptoHash, value: Arc<[u8]>) {
+        if let TrieCacheState::CachingChunk = self.cache_state.borrow().get() {
+            self.chunk_cache.borrow_mut().insert(key, value);
+        };
+    }
+
+    /// Get value for the given key from cache.
+    /// Return value if it is present and cost of retrieving the trie node corresponding to this key.
+    pub(crate) fn get_from_cache(&self, key: &CryptoHash) -> RawBytesWithCost {
+        match self.get_cache_position(key) {
+            // If no value is present, return None. If node for this key will be retrieved from DB, we have to charge
+            // full cost for it.
+            CachePosition::None => {
+                RawBytesWithCost { value: None, cost: TrieNodeRetrievalCost::Full }
+            }
+            // If value is present only in shard cache, we can copy value to the chunk cache. The cost must be full
+            // anyway because we didn't access the node in the chunk yet.
+            CachePosition::ShardCache(value) => {
+                self.try_put_to_chunk_cache(key.clone(), value.clone());
+                RawBytesWithCost { value: Some(value), cost: TrieNodeRetrievalCost::Full }
+            }
+            // If value is present in chunk cache, the cost is free.
+            CachePosition::ChunkCache(value) => {
+                RawBytesWithCost { value: Some(value), cost: TrieNodeRetrievalCost::Free }
+            }
+        }
+    }
+
+    /// Put a key-value pair into cache.
+    pub(crate) fn put_to_cache(&self, key: CryptoHash, value: Arc<[u8]>) {
+        self.shard_cache.put(key.clone(), value.clone());
+        self.try_put_to_chunk_cache(key, value.clone());
+    }
+
+    /// Set cache state.
+    pub fn set_state(&self, state: TrieCacheState) {
```

Comment:
I would probably implement this in the following manner: instead of set_state have a method switch_to_chunk_caching which returns RAII object switching the state back. It seems important to explicitly (on the caller function side) make sure that the state will be switched back to shard caching.

Reply:
Looks good. I just didn't understand how to make it work with the ownership structure we have (RuntimeExt -> TrieUpdate -> TrieStorage).

Will it be fine if I do it in a separate PR? I left a corresponding comment in `execute_function_call`.

Reply:
sg